### PR TITLE
Stop extracting mesh entities to the render world.

### PR DESCRIPTION
**Pull request by pcwalton**
This fixes a `FIXME` in `extract_meshes` and results in a performance improvement.

As a result of this change, meshes in the render world might not be attached to entities anymore. Therefore, the `entity` parameter to `RenderCommand::render()` is now wrapped in an `Option`. Most applications that use the render app's ECS can simply unwrap the `Option`.

Note that for now sprites, gizmos, and UI elements still use the render world as usual.

## Migration guide

* For efficiency reasons, some meshes in the render world may not have corresponding `Entity` IDs anymore. As a result, the `entity` parameter to `RenderCommand::render()` is now wrapped in an `Option`. Custom rendering code may need to be updated to handle the case in which no `Entity` exists for an object that is to be rendered.

### CODE REVIEW

File: crates/bevy_render/src/render_phase/draw.rs [hunk 1]
```diff
@@ -220,13 +225,22 @@ macro_rules! render_command_tuple_impl {
             fn render<'w>(
                 _item: &P,
                 ($($view,)*): ROQueryItem<'w, Self::ViewQuery>,
-                ($($entity,)*): ROQueryItem<'w, Self::ItemQuery>,
+                maybe_entities: Option<ROQueryItem<'w, Self::ItemQuery>>,
```
Comment (NthTensor) [inline crates/bevy_render/src/render_phase/draw.rs:228]: Is this correct? Why did you remove the macro decorations?

**Review by superdump — APPROVED** [review]

**Comment by superdump** [comment]
CI failed formatting.

**Review by NthTensor — APPROVED** [review]
Code looks good to me.

Profiling reveals a small but reliable improvement in `extract_meshs`, but I could not detect a change in the frame-rate or render schedule timings on the scenes I have available. Tested on `many_cubes`, `many_foxes` and several of the gltf test scenes using `scene_viewer`.

This is from `many_cubes`.

sc_1707534181

This PR paves the way for some other improvements that could take `extract_meshes` down into the microseconds most of the time.

The breaking change is pretty minor so I think we could still merge this for 0.13.

**Review by JMS55 — APPROVED** [review]

File: crates/bevy_pbr/src/render/mesh.rs [hunk 2]
```diff
@@ -260,8 +260,6 @@ pub struct RenderMeshInstances(EntityHashMap<Entity, RenderMeshInstance>);
 pub struct Mesh3d;
```
Comment (james7132) [inline crates/bevy_pbr/src/render/mesh.rs:260]: This might no longer be needed after this change.

**Review by james7132 — COMMENTED** [review]
The render command encoding seems to be ~4% slower on average (tested on `many_foxes`): 
image

This is probably a result of the panicking added into RenderCommand implementations, and this is likely still preferable than spending the time in the extract schedule as it blocks both the render and main schedules.

I can confirm similar performance improvements to what @NthTensor is noting with everything else though.

**Review by james7132 — APPROVED** [review]

**Comment by rparrett** [comment]
The migration guide here is a bit mysterious. Rust's suggestions for fixing up the function params seem good enough, but maybe it could be amended to specifically mention returning `RenderCommandResult::Failure`, or add a before/after code block.
